### DISCUSSION TITLE
Hebrew Diacritics Positioning Error In Word Processor

### DISCUSSION THREAD

**Discussion by JNord03** [fontforge/fontforge#5325] [General]
Hey,
I'm using FontForge to create a Hebrew font that supports advanced diacritics, i.e. distinguishes between sheva naḥ and dagesh ḥazaq/qal. The Unicode symbols for sheva and dagesh represent the latter, while an unassigned Unicode space is used for the former (I also expanded the glyphs' stroke width to make the distinction visible).
Within each set, both marks use the exact same anchors, so in theory, they should be positioned at the same place. The metrics window in FontForge shows what it's _supposed_ to look like:
<img width="531" alt="Metrics Window" src="https://github.com/fontforge/fontforge/assets/51972670/b449ffac-dc67-45ba-92d5-f4115d59e49e">
But my word processor (LibreOffice) misplaces the custom characters:
<img width="159" alt="Word Processor" src="https://github.com/fontforge/fontforge/assets/51972670/d435d542-e94b-4fc4-ab9c-9f05d2df2f60">
I already made sure the problem isn't caused by the increased stroke thickness, and all four glyphs are assigned to the "Mark" class in the glyph info. Moving the unassigned characters across Unicode blocks (from "Hebrew" to "Alphabetical Presentation Forms" or into the "Private Use Area") didn't help, either.

EDIT: Assigning these glyphs a location defined by Unicode as a Hebrew mark gets rid of the problem. But placing it in "Combining Diacritical Marks" doesn't. It seems like all code points, with a few exceptions, are hardcoded to be treated as a letter, not a mark, in LibreOffice. Any idea how that works and how to override it (as I don't want to use an "incorrect" Unicode character for my custom glyphs)?

**Comment by iorsh** [comment]
Can you try to display the glyphs in some simple application, like GEdit? This would more likely show you the Harfbuzz placement as-is, since the LibreOffice can sometimes meddle with it.

**Reply by JNord03** [reply]
Thank you Maxim! By the way, I'm a big fan of your Culmus fonts. I also hope to release this font on Open Siddur when it's ready.
I tried displaying the glyphs in Notepad++ and the same issue occurs there. I also noticed that the Meteg is placed incorrectly (I haven't fully figured out mark-to-mark placement yet - how do I make sure the Meteg is centered when used alone, but moved to the left when used with a niqqud mark?). Could you have a look at the file? The glyphs in question are mapped to U+05C8 and U+O5C9.

**Reply by iorsh** [reply]
Posting here excerpt from private communication:

> You can use David CLM (https://culmus.sourceforge.io/download.html) as a reference for meteg and yerushalam features. Note the order of the lookups, it's important.
> 
> Below I tried to summarize the method. I developed it a few years ago; hopefully won't miss any detail.
> 
> Meteg (aleph + shwa + meteg example):
> 
> 1. "CtxAnchors_M_D1" lookup is responsible to attach shwa to aleph not at the middle, but at the right. Note that it is not associated with any script, so it shouldn't work by itself
> 2. "ColumnToBaseMark" is responsible for attaching meteg to shwa (it's a mkmk type), and it's associated with Hebrew, so should work automatically whenever meteg comes next to shwa
> 3. "BaseDiac contextual 0" subtable is responsible to activate "CtxAnchors_M_D1" lookup whenever it encounters the sequence aleph + shwa + meteg, and it activates "CtxAnchors_M_D1" over shwa (causing it to attach to aleph on the side)
> 
> Yerushalam:
> 
> 1. "CtxAnchors_Yerushalam" is responsible for attaching hirik to lamed far at the left. It is not associated with any script, so it shouldn't work by itself
> 2. "BaseDiac contextual 22" subtable is responsible to activate "CtxAnchors_Yerushalam" lookup whenever it encounters the sequence lamed + patah + hirik, and it activates "CtxAnchors_Yerushalam" over hirik (causing it to attach to lamed as appropriate)
> 
> All the process above can be automated by scripts I wrote a while ago - see https://sourceforge.net/p/culmus/scripts/ci/master/tree/fontforge_python/
> Their application is not quite intuitive, but they save me a lot of work. You can try to put them in ~/.config/fontforge/python and they will appear in the FontForge "Tools" menu.
